### PR TITLE
Implemented rudimentary generics

### PR DESCRIPTION
**Pull request by Baudin999**
Change contains the following changed:

- Allow simple generic types
- Add a new `ConsumeWhile` which also has an `Or` in there
- Implemented `IsGeneric` property on an `ASTTypeDefinition`

### CODE REVIEW

File: Compiler/AST/ASTType.cs [hunk 1]
```diff
@@ -120,7 +120,7 @@ public static (List<ASTError>, ASTType) Parse(IParser parser, List<ASTAnnotation
 
         }
 
-        public object Clone()
+        public ASTType Clone()
```
Comment (lucashorward) [inline Compiler/AST/ASTType.cs]: Why not use the IClonable function?

File: Compiler/VisitorBase.cs [hunk 2]
```diff
@@ -91,11 +91,15 @@ private T d
         {
             get
             {
+#pragma warning disable CS8653 // A default expression introduces a null value for a type parameter.
```
Comment (lucashorward) [inline Compiler/VisitorBase.cs]: Is there no better way of resolving this?

File: Compiler/Resolver.cs [hunk 3]
```diff
@@ -98,11 +82,84 @@ private ASTTypeField PluckField(ASTPluckedField field)
...
+            }
+        }
+
+        private void ResolveType(ASTType t, List<IASTNode> nodes, List<ASTError> errors)
+        {
+            t.Extensions.ToList().ForEach(e =>
+            {
+                if (!(FindNode(e) is ASTType extendedFrom))
+                {
+                    // TODO: Handle error gracefully.
```
Comment (lucashorward) [inline Compiler/Resolver.cs]: TODO?

File: Compiler/AST/ASTType.cs (hunk 1 above)
Comment (Baudin999) [reply Compiler/AST/ASTType.cs]: Was implemented as a simple Clone. Still need to refactor that to the IClonable interface. Will be next time once we've revisited the cloning.

File: Compiler/VisitorBase.cs (hunk 2 above)
Comment (Baudin999) [reply Compiler/VisitorBase.cs]: Maybe, not really sure atm

**Pull request merged**